### PR TITLE
Project Dallas

### PR DESCRIPTION
**Pull request by nojaf**
The F# compiler syntax tree. It will always have a special place in my heart and all that but in a lot of cases, it is not the tree Fantomas truly needs to restore the source code.

A couple of shortcomings (from Fantomas' point of view) come to mind:
- Code comments are floating in the tree. We still need to do our own detection to which nodes the comment belongs. (the concept of Trivia)
- There is a lot of stuff in the current tree we don't really use. This is why we so heavily use partial active patterns (in `SourceParser.fs`) to get around that. Think of `DebugPoint`, `SynValData` and others.
- The way some nodes are grouped isn't typically useful for Fantomas. Recursive bindings or types for example. We consider the nested children to be on the same level, while they are not in the F# AST.
- Something joining two nodes is more convenient to work it. For example, named computation expressions.

These shortcomings lead to a more complex story than necessary for newcomers.
Our [guide on missing code comments](https://fsprojects.github.io/fantomas/docs/contributors/The%20Missing%20Comment.html) is a good example of that. You really need to jump through a lot of hoops to fix this problem.

So, after carefully considering this, we think the situation could be improved by having our own syntax tree. In this PR we are laying the foundation for this. I believe it will unlock some more opportunities in the future due to the new approach.

In short, the premise goes a bit like this:
- All new type nodes in the new Tree implement the `Node` interface. A `Node` can store `TriviaNode`s before and after. (See `SyntaxOak.fs`)
- In `Fangorn.fs` we map the F# compiler AST to this new Node model.
- In `Flowering.fs` we insert the found trivia (comments, newline, directives) into the existing tree. 
- And in `CodePrinter2.fs` we traverse the new tree to collect the `WriterEvents`

At the time of writing, 83% of the unit tests are passing. So, there still is some work left to do.
From a style point of view, we try to keep everything exactly as it was. But due to the size of this refactor, we don't necessarily guarantee that nothing will change. Of course, we will respect the style guide at all times. Some things might just change slightly due to convenience.

From an end-user point of view, this won't really affect the usage of the tool. Some performance will be gained, but I don't expect anything significant.

Some todo items:
- [x] Fix all remaining unit tests
- [x] Restore format selection
- [x] Verify everything works for existing codebases (Fantomas, FsAutocomplete, F# compiler to name a few)
- [ ] Rename the new files and type names.
- [x] Add signature files where it makes sense.
- [x] Update the online tool to have a viewer for the new tree.
- [x] Update documentation

### CODE REVIEW

**Comment by nojaf** [comment]
90% of `Fantomas.Core.Tests` are passing 🎉

**Comment by nojaf** [comment]
93%

**Comment by auduchinok** [comment]
> In Fangorn.fs we map the F# compiler AST to this new Node model.
In Flowering.fs we insert the found trivia (comments, newline, directives) into the existing tree.

The names are interesting, but maybe it would be better to use self-evident names to make it easier for other, especially new, contributors? 🙂

**Comment by nojaf** [comment]
@auduchinok yup, already part of my to-do list (see above) 😊.

**Comment by nojaf** [comment]
96%

**Comment by nojaf** [comment]
98%

**Comment by nojaf** [comment]
Good morning all 😊, I was able to restore all unit tests of Fantomas.Core!
Next, I started looking at formatting selections. I think we will be able to pull this off with the new tree. I have one example working, the rest is still a work in progress.
(So at [78bbac7](https://github.com/fsprojects/fantomas/pull/2626/commits/78bbac71d79276b5ddcac5a954ec4f44c1614256) not all tests of Fantomas.Core.Tests are passing, the `FormatSelectionOnly` will fail.)

I suspect I'll have the format selection wrapped up by the end of the week.

**Comment by nojaf** [comment]
Works for `Fantomas`, `FsAutocomplete` & `dotnet/fsharp`.

File: docs/docs/contributors/Transforming.md [hunk 1]
```diff
@@ -74,6 +74,31 @@ let a =
...
+
+The untyped syntax tree from the F# compiler is used as an intermediate representation of source code in the process of transforming a text file to binary.  
+The AST is optimized for the use-case of generating binary. What we try to do in Fantomas is stop at the first AST level and go back to source text.
+
+The F# compiler was never designed with our use-case in mind and yet it has served us very well for years. 
+In the past we did not have our own tree and were able to pull of formatting by traversing the compiler tree.
+This of course had its limitations and we had to overcome these with some hacks.
+
+Alas, some things in the AST aren't shaped the way we would like them to be. Sometimes, there is too much information, other times to little.
+To stream line our entire process, we've decide to map the untyped tree to our own custom object model.
```
Comment (dawedawe) [inline docs/docs/contributors/Transforming.md:87]:
```suggestion
To stream line our entire process, we've decided to map the untyped tree to our own custom object model.
```

File: docs/docs/contributors/Transforming.md [hunk 2]
```diff
@@ -74,6 +74,31 @@ let a =
...
+This introduces a lot of flexibility and simplifies our story.
+
+> I thought Fangorn was dangerous - Gimli, son of Glóin
+
+In `Fangorn.fs` we map the AST to our tree model. Some of the benefit we get out of this:
```
Comment (dawedawe) [inline docs/docs/contributors/Transforming.md:92]:
```suggestion
In `Fangorn.fs` we map the AST to our tree model. Some of the benefits we get out of this:
```

File: docs/docs/contributors/Transforming.md [hunk 3]
```diff
@@ -74,6 +74,31 @@ let a =
...
+
+- The Oak model does not differentiate between implementation files and signature files. We use one tree model which allow for optimal code re-use in `CodePrinter2.fs`.
```
Comment (dawedawe) [inline docs/docs/contributors/Transforming.md:94]:
```suggestion
- The Oak model does not differentiate between implementation files and signature files. We use one tree model which allows for optimal code re-use in `CodePrinter2.fs`.
```

File: docs/docs/contributors/Transforming.md [hunk 4]
```diff
@@ -74,6 +74,31 @@ let a =
...
+- We don't map all possible combinations of AST into our model. Sometimes valid AST code can in theory be created, 
+  but will in practise never exist. For example [SynTypeDefnRepr.Exception](../../reference/fsharp-compiler-syntax-syntypedefnrepr.html#Exception). It is defined in `SyntaxTree.fs` yet the parser (`pars.fs`) will never create it.
+  The F# compiler uses this later in the typed tree. We will throw an exception when encountering this during the mapping as we have the foresight of what the parser doesn't create.
+- Recursive types are all consider as toplevel types. This is not the case in the AST but we map it as such.
```
Comment (dawedawe) [inline docs/docs/contributors/Transforming.md:98]:
```suggestion
- Recursive types are all considered as toplevel types. This is not the case in the AST but we map it as such.
```

File: docs/docs/contributors/Transforming.md [hunk 5]
```diff
@@ -120,83 +145,20 @@ The AST does contain a node for the line comment, but we cannot restore it when
...
-    Range = [4,3--4,4]
-    AddBefore = true
-}
-```
-
-Lastly, all the instructions are stored in the `Context` record.  
+#### Inserting trivia
 
-To summarize, the `Context` initialization:
+Once we have the trivia, we can insert them to an `Node` they belong to.
```
Comment (dawedawe) [inline docs/docs/contributors/Transforming.md:154]:
```suggestion
Once we have the trivia, we can insert them to a `Node` they belong to.
```

File: docs/docs/contributors/Traverse.md [hunk 6]
```diff
@@ -5,16 +5,16 @@ index: 6
 ---
 # Fantomas.Core overview (2)
 
-Once the `Context` is constructed, we can traverse the syntax tree to capture all the `WriterEvent`s.
+Once the `Oak` and populated with all the found trivia, we can traverse the `Oak` to capture all the `WriterEvent`s.
```
Comment (dawedawe) [inline docs/docs/contributors/Traverse.md:8]:
```suggestion
Once the `Oak` is populated with all the found trivia, we can traverse the `Oak` to capture all the `WriterEvent`s.
```

**Review by dawedawe — APPROVED** [review]

**Pull request merged**